dbms/src/AggregateFunctions/AggregateFunctionWindowFunnel.h

```diff
@@ -215,19 +215,13 @@ class AggregateFunctionWindowFunnel final
 
     void add(AggregateDataPtr place, const IColumn ** columns, const size_t row_num, Arena *) const override
     {
-        UInt8 event_level = 0;
-        for (const auto i : ext::range(1, events_size + 1))
+        const auto timestamp = static_cast<const ColumnVector<UInt32> *>(columns[0])->getData()[row_num];
+        // reverse iteration and stable sorting are needed for events that are qualified by more than one condition.
+        for (auto i = events_size; i > 0; --i)
         {
             auto event = static_cast<const ColumnVector<UInt8> *>(columns[i])->getData()[row_num];
             if (event)
-            {
-                event_level = i;
-                break;
-            }
-        }
-        if (event_level)
-        {
-            this->data(place).add(static_cast<const ColumnVector<UInt32> *>(columns[0])->getData()[row_num], event_level);
+                this->data(place).add(timestamp, i);
```

Comment:
Thank you, but how to avoid duplicate match? 
Eg.
windowFunnel(4)(timestamp, event <= 2, event <=4, event <=6, event <=8)
There's just one row data with `event = 2` , but it could have the result 4.

Reply:
I got it, nice solution.